README.md
Smart Home Web
=======================

Este proyecto fué realizado para la asignatura Desarrollo de Aplicaciones de la Maestria en Internet de las Cosas de la Universidad de Buenos Aires. Se utilizó como base el proyecto [app-fullstack-base](https://github.com/gotoiot/app-fullstack-base).

![architecture](doc/home.png)

### Instalar las dependencias

- Docker
- Docker-compose

### Descargar el código

- Desde una consola:

```
git clone https://github.com/santiagosalamandri/app-fullstack-base.git
```
- Desde la [web](https://github.com/santiagosalamandri/app-fullstack-base).


### Ejecutar la aplicación

- Ejecutar la aplicación correr el comando `docker-compose up` desde la raíz del proyecto.
- Cliente web: [http://localhost:8000/](http://localhost:8000/) 
- Admin de la DB: [localhost:8001/](http://localhost:8001/). 


## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.


### Estructura de la DB

La base de datos se llama "smart_home" y pose la siguiente tabla "Device" con los siguiente campos:


| Nombre                   	| Tipo        	|
|--------------------------	|-------------	|
| id (Clave primaria)      	| int(11)     	|
| name                     	| varchar(64) 	|
| description  	            | varchar(128)	|
| state                    	| tinyint(1)  	|
| status                   	| int(11)     	|
| type                     	| int(11)     	|

## Detalles principales 🔍

- id: Clave primaria para determinar el dispositivo.
- name: Nombre del dispositivo.
- description: Descripción del dispositivo.
- type: Tipo de dispositivo. 
   - Tipo "0" para dispositivos "on/off".
   - Tipo "1" para dispositivos "dimerisables".
- state: Estado del dispositivo de tipo "0". 
    - "0" para apagado.
    - "1" para encendido.
- status: Estado del dispositvo de tipo "1".
    - Valores posibles entre 0 y 100.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas.

### Organización del proyecto


```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

- Para el frontend se modificaron los siguientes archivos:
    - devices.ts: Archivo con la definición de la clase Devices con las características de los dispositivos.
    - handlerInterface.ts: Interfaz a implementar para recibir las respuestas del backend.
    - main.ts: Programa principal donde se gestiona la lógica del frontend.
    - index.html : Archivo utilizado por el navegador para visualar la página web.
- Para el backend se modificaron los siguientes archivos:
    - index.js : Archivo donde se generaron las rutas para procesar las peticiones HTTP y la conexión con la base de datos

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.


### Agregar un dispositivo

- Hacer click en "Agregar dispositivo".
- Completar los datos del nuevo dispositivo (Nombre,Descripción y Tipo).
- Hacer click en "Confirmar".

### Eliminar un dispositivo

- Hacer click en "Eliminar" en la tarjeta que corresponda.

### Actualizar un dispositivo

- Hacer click en el switch o slider del dispositivo que se quiera cambiar su estado.

### Editar un dispositivo

- Hacer click en "Editar".
- Completar los datos del dispositivo (Nombre, Descripción y/o Tipo).
- Hacer click en "Confirmar".


### Frontend

El frontend esta conformado por 3 partes principales:
- Header: Incluye el nombre del proyecto y la ayuda
- Body: Donde se encuentran una matriz de tarjetas con los dispositivos y las acciones disponibles:
    - Un botón para agregar nuevos dispositivos.
    - Cada tarjeta incluye la informacion del dispositivo y la posibilidad de:
        - Cambiar su estado
        - Editar el dispositivo(Nombre,Descricción y Tipo).
        - Eliminar dispostivo.
    
- Footer: Contiene datos de contacto

### Backend

El frontend envía las peticiones HTTP a la backend, y este verifica los parametros. En caso de ser correctos,
realiza la consulta a la base de datos con la acción requerida.

#### Peticiones disponibles:

1) Devolver el estado de los dispositivos.

```json
{
    "url":"http://localhost:8000/devices/",
    "method": "get",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body": {
        "devices": [
            {
                "id": 1,
                "status": true,
                "description": "Kitchen light"
            }
        ]
    },
}
``` 

2) Devolver el estado de un dispositivo en particular.
```json
{
    "url":"http://localhost:8000/devices/:id",
    "method": "get",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body": {
        "devices": 
            {
                "id": 1,
                "status": true,
                "description": "Kitchen light"
            }
    },
}
``` 
3) Agregar un dispositivo.

```json
{
    "url":"http://localhost:8000/devices/",
    "method": "post",
    "request_headers": "application/json",
    "request_body": {
                    "name":"Light",
                    "description": "Kitchen light",
                    "type":0
                    },
    "response_code": 200,
    "request_body": {
        "devices": 
            {
                    "id": 1,
                    "name":"Light",
                    "description": "Kitchen light",
                    "type":0
            }
        
    },
}
``` 

4) Actualizar el estado de un dispositivo.

```json
{
    "url":"http://localhost:8000/devices/:id",
    "method": "post",
    "request_headers": "application/json",
    "request_body": 
                {
                "status": true,
                },
    "response_code": 200,
    "request_body": {
        "devices":{
                "id": 1,
                "status": true,
                }
    },
}
``` 

5) Actualizar los datos de un dispositivo.

```json
{
    "url":"http://localhost:8000/devices/:id",
    "method": "put",
    "request_headers": "application/json",
    "request_body": {
                    "name":"Light",
                    "description": "Kitchen light",
                    "type":0
                    },
    "response_code": 200,
    "request_body": {
        "devices": {
                    "id":1,
                    "name":"Light",
                    "description": "Kitchen light",
                    "type":0
                    }
    },
}
``` 
6) Borrar un dispositivo.

```json
{
    "url":"http://localhost:8000/devices/:id",
    "method": "delete",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body": {
        "devices":{
                "id": 1,
                }
    },
}
``` 

## Tecnologías utilizadas 🛠️


* [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
* [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
* [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
* [MySQL](https://www.mysql.com/) - Base de datos para consultar y almacenar datos.
* [PHPMyAdmin](https://www.phpmyadmin.net/) - Administrador web de base de datos.
* [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
* [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.



## Autor 👥


**[Santiago Salamandri](https://github.com/santiagosalamandri)**.



## Licencia 📄

Este proyecto está bajo Licencia ([MIT](https://choosealicense.com/licenses/mit/)). Podés ver el archivo [LICENSE.md](LICENSE.md) para más detalles sobre el uso de este material.


CHANGELOG.md
# DAW Base App - Changes Log

## Version 1.1

commit 950d8514caa050e5fd75a53a022b43fc8e4f2761
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 20:05:32 2021 +0100

    Arreglado valores por defecto en db.

commit 2cc574395e3a4633b73df867e2425e42c7760055
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 19:54:45 2021 +0100

    archivo user.js removido.

commit 58edfcde860c5dc8a35272d3070e9a52193eeaa7
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 19:43:10 2021 +0100

    Agregado cambio en la base de datos para que reconozca el campo status.

commit 7a42ca82060c8679a3f384edfda2f855f87c2d75
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 19:34:13 2021 +0100

    Arreglada interfaz de handlers y agregado fav.ico

## Version 1.0 

commit eb4f0564737cc1a88de82be3cf09db67d16bec9e
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 15:17:09 2021 +0100

    Refactor de codigo.

commit 2066e1caf1a3f4fb88adbba0429407968cd45124
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 14:56:33 2021 +0100

    Actualizado modal para edicion de dispositivo.

commit 05f8e34de18fee248345d97894abb006b052de1d
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 14:36:53 2021 +0100

    Integrado frontend con backend.

commit 2949b0fa6d9b5192ec7f203474fc2dcd53e9f40c
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 01:50:24 2021 +0100

    Agregada conexion con base de datos.

commit a015bdc290a074ea00e49035fa4983f85ab5e3e6
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Mon Jun 21 00:59:25 2021 +0100

    Actualizacion automatica cuando se cambian los controles de los dispositivos.

commit 066adf9e984099d06fff8c331e13cb731e65fc26
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sun Jun 20 23:40:24 2021 +0100

    Actualizadas acciones de frontend.

commit 5bc92ec18b7f984b5597de523a83ab212cba0deb
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sun Jun 20 20:28:26 2021 +0100

    Agregado agregar dispositivo y actualizan en Backend.

commit d31ea3f02628b2d25a236b859556ea60084cdedb
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sun Jun 20 19:30:37 2021 +0100

    Agregado Delete en backed

commit f95b7a7f3e77eba7f9b6dc35af4680b9af5681a3
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sun Jun 20 15:29:04 2021 +0100

    Agregado slider para devices tipo 1.

commit 7d4c63541e0485b7a0879a0670f049e3f2863c6c
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sat Jun 19 13:05:43 2021 +0100

    Agregado modal para edicion

commit 6dc960737db0421e0720c1a64b39041526b97f0e
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Sat Jun 19 11:16:21 2021 +0100

    Agregados botones de edicion y borrado de dispositivos.

commit ad1e8f8d58a83b9cb07d516d3b3844d7317e086c
Author: santiago salamandri <santiagosalamandri@gmail.com>
Date:   Fri Jun 18 21:14:24 2021 +0100

    Agregado getter para devices.



src/backend/index.js
//=======[ Settings, Imports & Data ]==========================================

var PORT = 3000;

var express = require('express');
var app = express();
var utils = require('./mysql-connector');

app.use(express.json());
app.use(express.static('/home/node/app/static/'));

const { json } = require('express');

/**
 * Devuelve un dispositivo especifico
 */
app.get('/devices/:id', function (req, res) {
    utils.query(`Select * from smart_home.Devices where id=${req.params.id}`, function (err, rta, field) {
        if (err) {
            res.send(err).status(400);
            return;
        }
        if (rta[0]) {
            res.send(rta[0]);
        }
        else {
            res.status(404).json({ "Error": "Not Found" });
        }
    });

});

/**
 * Actualizar estado de dispositivo
 */
app.post('/devices/:id', function (req, res) {
    let state = req.body.state? req.body.state :0;
    let status = req.body.status?req.body.status:0;
    
    utils.query(`UPDATE smart_home.Devices SET status='${status}', state='${state}' where id=${req.params.id}`, function (err, rta, field) {   //update field
        if (err) {
            res.send(err).status(400);
            return;
        }
        let result=rta.affectedRows;
        if(result ==1){
            utils.query(`SELECT * FROM smart_home.Devices WHERE id=${req.params.id} LIMIT 0,1`, function (err, rta, field) { //get one device
                if (err) {
                    res.send(err).status(400);
                    return;
                }
                res.send(rta[0]);
            });
        }
        else{
            res.status(404).json({ "Error": "Not Found" });
    
        }});
});
/**
 * Borrar dispositivo
 */
app.delete('/devices/:id', function (req, res) {
    utils.query(`DELETE FROM smart_home.Devices WHERE id=${req.params.id}`, function (err, rta, field) { //delete
        if (err) {
            res.send(err).status(400);
            return;
        }
        let result=rta.affectedRows;
        if(result ==1){
            res.status(200).json({ "id": req.params.id });

        }
        else{
            res.status(404).json({ "Error": "Not Found" });
 
        }
    });
});
/**
 * Actualizador datos de dispositivo
 */
app.put('/devices/:id', function (req, res) {
utils.query(`UPDATE smart_home.Devices SET name='${req.body.name}', description='${req.body.description}',type='${req.body.type}' where id=${req.params.id}`, function (err, rta, field) {   //update field
    if (err) {
        res.send(err).status(400);
        return;
    }
    let result=rta.affectedRows;
    if(result ==1){
        utils.query(`SELECT * FROM smart_home.Devices WHERE id=${req.params.id} LIMIT 0,1`, function (err, rta, field) { //get one device
            if (err) {
                res.send(err).status(400);
                return;
            }
            res.send(rta[0]);
        });
    }
    else{
        res.status(404).json({ "Error": "Not Found" });

    }});

});
/**
 * Agregar nuevo dispositivo
 */
app.post('/devices/', function (req, res) {
    if (req.body.name != undefined && req.body.description != undefined && req.body.type != undefined) {
        utils.query(`INSERT INTO smart_home.Devices(name,description,type) VALUES ('${req.body.name}','${req.body.description}',${req.body.type})`, function (err, rta, field) { //insert            if (err) {
            if (err) {
                res.send(err).status(400);
                return;
            }
            let index = rta.insertId;
            utils.query(`SELECT * FROM smart_home.Devices WHERE id=${index} LIMIT 0,1`, function (err, rta, field) { //get one device
                res.send(rta[0]);
            });

        });
    }
    else {
        res.status(400).json({ "Error": "Formato incorrecto" });
    }

});
/**
 * Listar todos los dispositivos
 */
app.get('/devices/', function (req, res, next) {
    let id = 11;
    let name = "Lámpara 2999";
    let type = 1;
    let description = "nueva lamp";
    utils.query(`SELECT * FROM smart_home.Devices`, function (err, rta, field) { 
        if (err) {
            res.send(err).status(400);
            return;
        }
        res.send(rta);
    });
});

//=======[ Main module code ]==================================================
app.listen(PORT, function (req, res) {
    console.log("NodeJS API running correctly");
});

//=======[ End of file ]=======================================================

src/frontend/ts/MyFramework.ts
class MyFramework {
  public getElementById(id: string): HTMLElement {
    return document.getElementById(id);
  }

  public requestPOST(url: string, response: HandlerHttpEvents, datos: any) {
    let xlm: XMLHttpRequest = new XMLHttpRequest();

    xlm.onreadystatechange = () => {
      if (xlm.readyState == 4) {
        response.responsePost(xlm.status, xlm.responseText);
      }
    }
    xlm.open("POST", url, true);
    xlm.setRequestHeader("Content-Type", "application/json");
    xlm.send(JSON.stringify(datos));
  }
  /**
   * requestGET
   */
  public requestGET(url: string, response: HandlerHttpEvents) {
    let xhr: XMLHttpRequest = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState == 4) {
        console.log("xhr.status: " + xhr.status);
        response.responseGet(xhr.status, xhr.responseText);
      }
    }
    xhr.open("GET", url, true)
    xhr.send();
    console.log("Ya hice el request!!")
  }
  public requestDELETE(url: string, response: HandlerHttpEvents) {
    let xhr: XMLHttpRequest = new XMLHttpRequest();
    xhr.onreadystatechange = () => {
      if (xhr.readyState == 4) {
        console.log("xhr.status: " + xhr.status);
        response.responseDelete(xhr.status, xhr.responseText);
      }
    }
    xhr.open("DELETE", url, true)
    xhr.send();
    console.log("Ya hice el request!!")
  }
  public requestPUT(url: string, response: HandlerHttpEvents, datos: any) {
    let xlm: XMLHttpRequest = new XMLHttpRequest();

    xlm.onreadystatechange = () => {
      if (xlm.readyState == 4) {
        response.responsePut(xlm.status, xlm.responseText);
      }
    }
    xlm.open("PUT", url, true);
    xlm.setRequestHeader("Content-Type", "application/json");
    xlm.send(JSON.stringify(datos));
  }
}

src/frontend/ts/main.ts
const BASE_URL = "http://localhost:8000/devices/";

class Main implements EventListenerObject, HandlerHttpEvents {
    public myFramework: MyFramework;
    public myDevices: Array<Device>;
    public deviceChanged: Device;
    public lastIdClicked: number;
    public main(): void {
        this.myFramework = new MyFramework();

    }
    public getDeviceById(id: number): Device {
        for (let device of this.myDevices) {
            if (device.id == id) {
                console.log("Found device: ", device.name);
                return device
            }
        }
        console.log("NOT FOUND!");
        return null; //Not found
    }
    public removeDeviceById(id: number): Array<Device> {
        let newArray: Array<Device> = this.myDevices.filter(item => item.id != id);
        return newArray;
    }
    public printDevices() {
        for (let device of this.myDevices) {
            console.log(device.id + " :" + device.name + " Desc:" + device.description + " Status:" + device.status + " State:" + device.state)
        }
    }
    public listDevices(devices: Array<Device>) {
        let showSwitch = "";
        let showSlider = "";
        let listaDisp = this.myFramework.getElementById("listaDisp");
        listaDisp.innerHTML = "";

        for (let disp of devices) {
            if (disp.type == 0) {
                showSwitch = "block";
                showSlider = "none";
            }
            else {
                showSwitch = "none";
                showSlider = "block";
            }
            let listaDisp = this.myFramework.getElementById("listaDisp");
            listaDisp.innerHTML += `<div class="col l4 s4 m4">
                                        <div class="card blue-grey darken-1" >
                                            <div class="card-content white-text" >
                                                <span class=card-title> ${disp.name} </span>
                                                    <p> ${disp.description}.</p>            
                                                    <br />
                                                    <div class="switch" style="display:${showSwitch}">
                                                        <label>
                                                        Off
                                                        <input id="disp_${disp.id}" type="checkbox"  >
                                                        <span class="lever"></span>
                                                        On
                                                        </label>
                                                    </div>
                                                    <div class="range-field" style="display:${showSlider}" >
                                                        <input type="range" id="range_${disp.id}"  min="0" max="100" />
                                                        </div>      
                                            </div>
                                            <div class="card-action">
                                                <a id=edit_${disp.id} class="waves-effect waves-light btn " href="#modal">Editar</a>
                                                <a id=del_${disp.id} class="waves-effect waves-light btn" href=#modal">Eliminar</a>
                                            </div>
                                        </div>
                                    </div>`;
        }
    }
    public updateDeviceStatus(response: string): void {
        let deviceUpdated = JSON.parse(response);
        let datos = this.myDevices.filter(item => item.id == deviceUpdated.id);
        if (datos.length > 0) {
            datos[0].name = deviceUpdated.name;
            datos[0].description = deviceUpdated.description;
            datos[0].type = deviceUpdated.type;
            datos[0].state = deviceUpdated.state;
            datos[0].status = deviceUpdated.status;
        }
        else {   //nuevo dispositivo
            this.myDevices.push(deviceUpdated);
        }
        this.responseGet(200, JSON.stringify(this.myDevices));
    }
    public populateForm(id: number) {
        let device = this.getDeviceById(id);

        const headerId = this.myFramework.getElementById('headerDisp');
        headerId.innerText = "Editar dispositivo";

        const nameId = this.myFramework.getElementById('formName');
        let name: HTMLInputElement = <HTMLInputElement>nameId;
        name.value = device.name;

        const formId = this.myFramework.getElementById('formDescription');
        let desc: HTMLInputElement = <HTMLInputElement>formId;
        desc.value = device.description;

        if (device.type == 0) {
            const typeOnOffId = this.myFramework.getElementById('typeOnOff');
            let typeOnOff: HTMLInputElement = <HTMLInputElement>typeOnOffId;
            typeOnOff.checked = true;
        }
        else {
            const typeDimId = this.myFramework.getElementById('typeDim');
            let typeDim: HTMLInputElement = <HTMLInputElement>typeDimId;
            typeDim.checked = true;
        }

        const elem = this.myFramework.getElementById('modal');
        let instance = window.M.Modal.getInstance(elem);
        instance.open();
    }
    public getValuesFromForm() {

        const nameId = this.myFramework.getElementById('formName');
        let name: HTMLInputElement = <HTMLInputElement>nameId;

        const formId = this.myFramework.getElementById('formDescription');
        let desc: HTMLInputElement = <HTMLInputElement>formId;

        const typeOnOffId = this.myFramework.getElementById('typeOnOff');
        let typeOnOff: HTMLInputElement = <HTMLInputElement>typeOnOffId;

        const typeDimId = this.myFramework.getElementById('typeDim');
        let typeDim: HTMLInputElement = <HTMLInputElement>typeDimId;

        let type = typeOnOff.checked ? 0 : 1;

        return { "id": this.lastIdClicked, "name": name.value, "description": desc.value, "type": type };
    }
    public actualizarModal() {
        const headerId = this.myFramework.getElementById('headerDisp');
        headerId.innerText = "Nuevo dispositivo";

        const formNameId = this.myFramework.getElementById('formName');
        let formName: HTMLInputElement = <HTMLInputElement>formNameId;
        formName.value = "";

        const formId = this.myFramework.getElementById('formDescription');
        let formDesc: HTMLInputElement = <HTMLInputElement>formId;
        formDesc.value = "";

        const typeOnOffId = this.myFramework.getElementById('typeOnOff');
        let typeOnOff: HTMLInputElement = <HTMLInputElement>typeOnOffId;
        typeOnOff.checked = true;
    }
    public handleEvent(ev: Event) {

        let objetoClick: HTMLElement = <HTMLElement>ev.target;
        switch (objetoClick.textContent) {
            case "Editar":
                let str = objetoClick.id.replace("edit_", "");
                this.lastIdClicked = parseInt(str);
                this.populateForm(parseInt(str));
                break;
            case "Eliminar":
                let strId = objetoClick.id.replace("del_", "");
                this.myFramework.requestDELETE(`${BASE_URL}${strId}`, this);
                break;
            case "Confirmar":
                if (this.lastIdClicked != 0) {
                    this.myFramework.requestPUT(`${BASE_URL}${this.lastIdClicked}`, this, this.getValuesFromForm());
                } else {
                    this.myFramework.requestPOST(`${BASE_URL}`, this, this.getValuesFromForm());
                }
                this.lastIdClicked;
                this.actualizarModal();
                break;
            case "Descartar":
                this.actualizarModal();
                break;
            default:    //Checkbox clicked
                let checkBox: HTMLInputElement = <HTMLInputElement>objetoClick;
                var res = objetoClick.id.substring(0, 6);
                let datos = {};
                let idDevice = "";
                if (res == "range_") {
                    idDevice = checkBox.id.replace("range_", "");
                    datos = { "status": checkBox.value }
                }
                else {
                    idDevice = checkBox.id.replace("disp_", "");
                    let value = checkBox.checked ? 1 : 0;
                    datos = { "state": value };
                }
                let id = parseInt(idDevice);
                this.myFramework.requestPOST(`${BASE_URL}${id}`, this, datos);
                break;
        }
    }

    responsePost(status: number, response: string) {
        console.log("Response POST: " + response);
        if (status == 200) {
            this.updateDeviceStatus(response);
        }
        else {
            alert("Request fallida");
        }
    }
    responseGet(status: number, response: string) {
        if (status == 200) {
            this.myDevices = JSON.parse(response);
            this.listDevices(this.myDevices);
            for (let disp of this.myDevices) {
                let checkDisp = this.myFramework.getElementById("disp_" + disp.id);
                checkDisp.addEventListener("click", this);
                let rangeSwitch = this.myFramework.getElementById("range_" + disp.id);
                rangeSwitch.addEventListener("click", this);
                let checkEdit = this.myFramework.getElementById("edit_" + disp.id);
                checkEdit.addEventListener("click", this);
                let checkDel = this.myFramework.getElementById("del_" + disp.id);
                checkDel.addEventListener("click", this);
                if (disp.type == 1) {
                    let update = <HTMLInputElement>this.myFramework.getElementById("range_" + disp.id);
                    update.value = ""+disp.status;
                }
                else {
                    let update = <HTMLInputElement>this.myFramework.getElementById("disp_" + disp.id);
                    update.checked = disp.state;

                }
            }
        }
        else {
            alert("Request fallida");
        }

    }
    responseDelete(status: number, response: string) {
        if (status == 200) {
            let id = JSON.parse(response);
            this.responseGet(200, JSON.stringify(this.removeDeviceById(id.id)));
        }
        else {
            alert("Request fallida");
        }

    }
    responsePut(status: number, response: string) {
        if (status == 200) {
            this.updateDeviceStatus(response);
        }
        else {
            alert("Request fallida");
        }

    }

}
window.addEventListener("load", () => {
    let miObjMain: Main = new Main();
    miObjMain.main();
    //Traigo los datos de dispositivos cuando la pagina termino de cargar
    miObjMain.myFramework.requestGET(`${BASE_URL}`, miObjMain);
    
    miObjMain.lastIdClicked = 0;

    //Registro modal y sus botones
    const elem = miObjMain.myFramework.getElementById('modal');
    const instance = this.M.Modal.init(elem, { dismissible: false });
    let confirmarId = miObjMain.myFramework.getElementById("confirmarId");
    confirmarId.addEventListener("click", miObjMain);
    let descartarId = miObjMain.myFramework.getElementById("descartarId");
    descartarId.addEventListener("click", miObjMain);
});